app/libs/deployments/google_play_store/release.rb

```diff
@@ -0,0 +1,135 @@
+module Deployments
+  module GooglePlayStore
+    class Release
+      include Loggable
+
+      def self.kickoff!(deployment_run)
+        new(deployment_run).kickoff!
+      end
+
+      def self.upload!(deployment_run)
+        new(deployment_run).upload!
+      end
+
+      def self.start_release!(deployment_run)
+        new(deployment_run).start_release!
+      end
+
+      def self.rollout_release!(deployment_run, rollout_value:, &blk)
+        new(deployment_run).rollout_release!(rollout_value:, &blk)
+      end
+
+      def self.halt_release!(deployment_run, &blk)
+        new(deployment_run).halt_release!(&blk)
+      end
+
+      def self.release_to_all!(deployment_run, &blk)
+        new(deployment_run).release_to_all!(&blk)
+      end
+
+      def initialize(deployment_run)
+        @deployment_run = deployment_run
+      end
+
+      attr_reader :deployment_run
+      alias_method :run, :deployment_run
+      delegate :provider,
+        :deployment_channel,
+        :build_number,
+        :release_version,
+        :staged_rollout?,
+        :google_play_store_integration?,
+        :staged_rollout_config,
+        :promotable?,
+        :release,
+        to: :run
+
+      def kickoff!
+        return run.upload! if run.step_run.similar_deployment_runs_for(run).any?(&:has_uploaded?)
+        Deployments::GooglePlayStore::Upload.perform_later(run.id)
+      end
+
+      def upload!
+        return unless google_play_store_integration?
+
+        run.with_lock do
+          return if run.uploaded?
+
+          run.build_artifact.with_open do |file|
+            result = provider.upload(file)
+            if result.ok?
+              run.upload!
+            else
+              run.fail_with_error(result.error)
+            end
+          end
+        end
+      end
+
+      def start_release!
+        return unless google_play_store_integration?
+
+        if staged_rollout?
+          run.engage_release!
+          rollout!
+        else
+          fully_release!
+        end
+      end
+
+      def halt_release!
+        return unless google_play_store_integration?
+
+        release.with_lock do
+          return unless run.rollout_started?
+          yield provider.halt_release(deployment_channel, build_number, release_version, run.staged_rollout.last_rollout_percentage)
+        end
+      end
+
+      def release_to_all!(&blk)
+        return unless google_play_store_integration?
+
+        release_with(rollout_value: Deployment::FULL_ROLLOUT_VALUE, &blk)
+      end
+
+      # TODO: handle known errors gracefully and show to users
```

Comment:
Is this `TODO` still necessary?

Reply:
the graceful handling is happening, but we are still not showing it to users.

I wanted to add a small UI change to show the human-readable `failure_reason` for when a deployment fails whether via manual trigger or a job.